<launch>

    <arg name="bag_file_use" default="false" />
    <arg name="config_file" default="$(find mloam)/config/config_realvehicle_mltest.yaml" />
    <arg name="data_source" default="pcd" />
    <arg name="data_path" default="/Monster/dataset/UDI_dataset/xfl_20200301/whole_travel/" />

    <arg name="run_mloam" default="true" />
    <arg name="run_aloam" default="false" />
    <arg name="run_legoloam" default="false" />

    <arg name="run_mloam_map_with_up" default="true" /> <!-- uncertainty propagation -->    
    <arg name="mloam_result_save" default="1"/>
    <arg name="mloam_start_idx" default="2100"/>
    <arg name="mloam_end_idx" default="25229"/>
    <arg name="mloam_delta_idx" default="2" />
    <arg name="mloam_time_now" default="0" />

    <arg name="output_path" default="/home/jjiao/trajectory_results/real_vehicle/rv08_2100_25229_test/rv08_mltest_top/" />
    <param name="use_sim_time" value="false" />

    <group if="$(arg run_mloam)">
        <node pkg="mloam" type="mloam_node_rv" name="mloam_node_rv" 
            args="$(arg config_file) 
                  $(arg data_source)
                  $(arg data_path)
                  $(arg mloam_result_save) 
                  $(arg output_path)
                  $(arg mloam_start_idx)
                  $(arg mloam_end_idx)
                  $(arg mloam_delta_idx)
                  $(arg mloam_time_now)" output="screen">
            <remap from="/base_pose_gt" to="/current_odom"/>
        </node>

        <group if="$(arg run_mloam_map_with_up)">
            <node pkg="mloam" type="lidar_mapper" name="lidar_mapper" 
                args="$(arg config_file) $(arg mloam_result_save) $(arg output_path) 1" output="screen"/>
        </group>
        <group unless="$(arg run_mloam_map_with_up)">
            <node pkg="mloam" type="lidar_mapper" name="lidar_mapper" 
                args="$(arg config_file) $(arg mloam_result_save) $(arg output_path) 0" output="screen"/>        
        </group>
    </group>

    <!-- publish static tf -->
    <!-- x y z yaw pitch roll -->
    <!-- x y z qx qy qz qw -->
    <include file="$(find rviz_car_model)/launch/default.launch" />
    <node name="static_tf_base_sensor_into_base_link" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 1 aft_mapped base_link 10"/>                
    <node name="static_tf_world_base_into_world_sensor" pkg="tf" type="static_transform_publisher"
        args="0 0.266 0.734 -0.216 0 0 0.976 world_base world 10"/>
    <node name="static_tf_world_sensor_into_base_sensor" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 1 world laser_0 10"/>

    <!-- merge cloud -->
    <group unless="$(arg run_mloam)">
        <node pkg="mloam" type="test_merge_pointcloud_rv" name="test_merge_pointcloud_rv" 
            args="1 $(arg data_path) $(arg mloam_start_idx) $(arg mloam_end_idx) $(arg mloam_delta_idx)"/>      
        <node name="static_tf_world_sensor_into_camera" pkg="tf" type="static_transform_publisher"
            args="0 0 0 0 0 0 1 world camera_init 10"/>            
    </group>

    <!-- run aloam  -->
    <group if="$(arg run_aloam)">
        <include file="$(find aloam_velodyne)/launch/aloam_exp.launch">
            <arg name="aloam_odom_path" default="/home/jjiao/trajectory_results/real_vehicle/rv01_20200301_whole_travel_test_2100_25229/stamped_aloam_odom_estimate.txt"/>
            <arg name="aloam_map_path" default="/home/jjiao/trajectory_results/real_vehicle/rv01_20200301_whole_travel_test_2100_25229/stamped_aloam_map_estimate.txt"/>
        </include>       
`  </group>        

    <!-- run legoloam  -->
    <group if="$(arg run_legoloam)">
        <include file="$(find lego_loam)/launch/legoloam_exp.launch">
            <arg name="lego_loam_odom_path" default="/home/jjiao/trajectory_results/real_vehicle/rv01_20200301_whole_travel_test_2100_25229/stamped_legoloam_odom_estimate.txt"/>
            <arg name="lego_loam_map_path" default="/home/jjiao/trajectory_results/real_vehicle/rv01_20200301_whole_travel_test_2100_25229/stamped_legoloam_map_estimate.txt"/>
        </include>
    </group>  

 </launch>
